### PR TITLE
Adds comment support

### PR DESCRIPTION
**Pull request by tdurieux**

### CODE REVIEW

**Comment by seintur** [comment]
Buggy copied/pasted comment in CtVisitor.java

	/**
	 * Visits an access to a super invocation.
	 */
	void visitCtComment(CtComment comment);

**Comment by monperrus** [comment]
looks promising!

File: src/main/java/spoon/compiler/Environment.java [hunk 1]
```diff
@@ -297,6 +297,16 @@ void report(Processor<?> processor, Level level,
 	void setGenerateJavadoc(boolean generateJavadoc);
 
 	/**
+	 * Returns the value of the option generate-comments.
+	 */
+	boolean isGenerateComments();
```
Comment (monperrus) [inline src/main/java/spoon/compiler/Environment.java]: keepInlineComments

File: src/main/java/spoon/compiler/Environment.java [hunk 2]
```diff
@@ -297,6 +297,16 @@ void report(Processor<?> processor, Level level,
...
+
+	/**
+	 * Sets the option generate-comments to generate comments of the project on the source generated.
+	 */
+	void setGenerateComments(boolean generateComments);
```
Comment (monperrus) [inline src/main/java/spoon/compiler/Environment.java]: setInlineCommentSupport

File: src/main/java/spoon/reflect/code/CtComment.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ */
+package spoon.reflect.code;
+
+import spoon.reflect.declaration.CtElement;
+
+/**
+ * This code element defines a comment
+ *
+ */
+public interface CtComment extends CtElement, CtStatement {
```
Comment (monperrus) [inline src/main/java/spoon/reflect/code/CtComment.java]: only extends CtStatement

File: src/main/java/spoon/reflect/code/CtComment.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	String getContent();
+
+	boolean isInline();
```
Comment (monperrus) [inline src/main/java/spoon/reflect/code/CtComment.java]: CtCommentType getCommentType()
 enum CtCommentType JAVADOC /**, ENCLOSED /*, LINE //

**Comment by monperrus** [comment]
add Javadoc DocComment in List<Comment> (and remove the original String field in CtElement)

**Comment by monperrus** [comment]
reuse same option as javadoc (and change doc of this doc option)

**Comment by tdurieux** [comment]
Done

File: src/main/java/spoon/reflect/code/CtComment.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	 */
+	String getContent();
+
+	<E extends CtComment> E setContent(String content);
+
+	/**
+	 * Get the type of the comment
+ 	 * @return the comment type
+	 */
+	CommentType getType();
```
Comment (monperrus) [inline src/main/java/spoon/reflect/code/CtComment.java]: getCommentType (in order not to confuse with getType of CtTypedElem

File: src/main/java/spoon/reflect/declaration/CtElement.java [hunk 6]
```diff
@@ -76,6 +77,7 @@
 	 * Returns the text of the documentation ("javadoc") comment of this
 	 * element.
 	 */
+	@Deprecated
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtElement.java]: I would not deprecate them (we only unify the backend data structure)

File: src/main/java/spoon/reflect/declaration/CtElement.java [hunk 7]
```diff
@@ -253,4 +256,16 @@
 	 * Returns the metadata keys stored in an element.
 	 */
 	Set<String> getMetadataKeys();
+
+	<E extends CtElement> E setComments(List<CtComment> comments);
+
+	/**
+	 * The list of comments
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtElement.java]: give examples?

File: src/main/java/spoon/reflect/declaration/CtElement.java [hunk 8]
```diff
@@ -253,4 +256,16 @@
...
+	 * @return the list of comment
+	 */
+	List<CtComment> getComments();
+
+	<E extends CtElement> E addComment(CtComment comment);
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtElement.java]: javadoc?

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 9]
```diff
@@ -239,7 +245,7 @@ public SourcePosition getPosition() {
 	}
 
 	@Override
-	public final int hashCode() {
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: keep this final

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 10]
```diff
@@ -64,9 +64,11 @@
 	protected static final Logger LOGGER = Logger.getLogger(CtElementImpl.class);
 	public static final String ERROR_MESSAGE_TO_STRING = "Error in printing the node. One parent isn't initialized!";
 
+	private List<CtComment> comments = new ArrayList<CtComment>();
+
 	// we don't use Collections.unmodifiableList and Collections.unmodifiableSet
 	// because we need clear() for all set* methods
-	// and UnmodifiableList and unmodifiableCollection are not overridable (not visible grrrr)
+	// and UnmodifiableList and unmodifiableCollection are not override (not visible grrrr)
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: "overridable" was correct, right?

File: src/main/java/spoon/reflect/code/CtComment.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ */
+package spoon.reflect.code;
+
+/**
+ * This code element defines a comment
+ *
+ */
+public interface CtComment extends CtStatement {
+	enum CommentType {
+		FILE_COMMENT, // comment at the top of a file
```
Comment (monperrus) [inline src/main/java/spoon/reflect/code/CtComment.java]: FILE before the package line (typically the license)

File: src/main/java/spoon/reflect/code/CtComment.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+		JAVADOC,      // /** */
```
Comment (monperrus) [inline src/main/java/spoon/reflect/code/CtComment.java]: before methods, fields, types

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 13]
```diff
@@ -271,8 +277,15 @@ public boolean removeAnnotation(CtAnnotation<? extends Annotation> annotation) {
 		return (List<?>) annotations != (List<?>) emptyList() && this.annotations.remove(annotation);
 	}
 
+	@Deprecated
 	public <E extends CtElement> E setDocComment(String docComment) {
-		this.docComment = docComment;
+		for (CtComment ctComment : comments) {
+			if (ctComment.getCommentType() == CtComment.CommentType.JAVADOC) {
+				ctComment.setContent(docComment);
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: documented in the interface

**Comment by monperrus** [comment]
doc of comment type enum, depreciation, documentation of contract for java doc

File: .travis.yml [hunk 14]
```diff
@@ -1,5 +1,7 @@
 language: java
 
+sudo: true
```
Comment (GerardPaligot) [inline .travis.yml]: Can you change to `sudo: required` ? https://docs.travis-ci.com/user/ci-environment/#Virtualization-environments

**Comment by tdurieux** [comment]
Done

File: src/main/java/spoon/support/reflect/eval/VisitorPartialEvaluator.java [hunk 15]
```diff
@@ -743,6 +744,11 @@ public void visitCtCircularTypeReference(CtCircularTypeReference reference) {
 	}
 
 	@Override
+	public void visitCtComment(CtComment comment) {
+		throw new RuntimeException("Unknow Element");
```
Comment (seintur) [inline src/main/java/spoon/support/reflect/eval/VisitorPartialEvaluator.java]: Typo: Unknow -> Unknown

File: src/main/java/spoon/support/reflect/code/CtCommentImpl.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	@Override
+	public void accept(CtVisitor visitor) {
+		visitor.visitCtComment(this);
+	}
+
+	/**
+	 * The comments are not printed during the CtElement equality.
+	 * The method are thus override for CtComment
```
Comment (seintur) [inline src/main/java/spoon/support/reflect/code/CtCommentImpl.java]: Language fixing: The method is this overridden for CtComment.

File: src/main/java/spoon/support/compiler/jdt/JDTCommentBuilder.java [hunk 17]
```diff
@@ -0,0 +1,458 @@
...
+		}
+		// adds the comment to the nearest element
+		if (best != null) {
+			best.addComment(comment);
+		}
+		return best;
+	}
+
+	/**
+	 * Inserts the comment into in AST
```
Comment (seintur) [inline src/main/java/spoon/support/compiler/jdt/JDTCommentBuilder.java]: Typo: Inserts the comment into the AST.

**Pull request merged**